### ISSUE TITLE
GraphRenderer breaks custom Hover callbacks

### ISSUE THREAD

**Issue by felixpatzelt** [bokeh/bokeh#8347]
Hi, using CustomJS callbacks in the HoverTool appears to be broken with the GraphRenderer. The callback is never executed and `TypeError: undefined is not an object (evaluating 'n.data_source.inspected')` is printed to the browser's console. I didn't see this issue with other renderers.

I'm using  bokeh 0.13.0 and networkx 2.1 with Python 3.6.6 (installed by Anaconda) and Jupyter 4.4.0 on macOS 10.13.5.

#### Minimal example
```Python
from bokeh.io import show, output_notebook
from bokeh.models import Range1d, Plot, CustomJS, Circle, HoverTool
from bokeh.models.graphs import from_networkx
import networkx as nx

output_notebook()

# create the graph renderer
G = nx.Graph()
G.add_weighted_edges_from([(1, 2, 0.125), (1, 3, 0.75), (2, 4, 1.2), (3, 4, 0.375)])
graph = from_networkx(G, nx.layout.circular_layout, scale=1.8, center=(0,0))

# create a circle just to also test another renderer
c = Circle()
c.x = 0
c.y = 0
c.radius = .1

# build plot
plot  = Plot(x_range=Range1d(-3, 3), y_range=Range1d(-3, 3))
plot.renderers.append(graph)
plot.add_glyph(c)

# this just prints an error to the console: 
# "TypeError: undefined is not an object (evaluating 'n.data_source.inspected')"
hover_tool = HoverTool(
    tooltips="index: @index", 
    callback=CustomJS(code='console.log("foo");'), 
    renderers=[plot.renderers[0]]
)

# this prints "foo" and no error to the console, 
# but it also doesn't allow us to interact with the graph
#hover_tool = HoverTool(
#    tooltips="index: @index", 
#    callback=CustomJS(code='console.log("foo");'), 
#    renderers=[plot.renderers[1]]
#)

plot.add_tools(hover_tool)

show(plot)
```

**Comment by bryevdv** [comment]
@felixpatzelt The `GraphRenderer` is actually a composite of two other renderers, `graph.node_renderer` and `graph.edge_renderer`.  The graph renderer itself does not have any data source, only the node and edge renderers do, which is the proximate cause of the message. You need to set one one of those in the `renderers` list for the hover tool, depending on whether you want the tool to hover for nodes or edges.

@canavandl is this documented anywhere? Perhaps a validation check could warn about this? Or we could just remove `GraphRenderer` instances from `renderers` automatically since they never make sense to be configured on a hover tool.

**Comment by bryevdv** [comment]
@felixpatzelt Yes, that's why I suggested we might just decide to remove `GraphRenderer` instances automatically. Specifically I think Bokeh should: 

* silently exclude them from being added on the JS side when no renderers are given 
* issue a valuation error if they explicitly added on the Python side 

Alternatively to the second bullet, we could potentially refine the type system to prevent `GraphRenderer` from being added explicitly on the python side at all.

**Comment by meenurajapandian** [comment]
Even after explicitly mentioning the renderers in HoverTool (` renderers=[graph.edge_renderer, graph.node_renderer] '), it doesn't work. In fact, when no renderer is mentioned, the hover details are displayed for both nodes and edges. When explicitly mentioned, no details are displayed.

**Comment by bryevdv** [comment]
@meenurajapandian please provide actual code and relevant version information. It is not valuable for us to speculate about the details of what you are doing.
